File: core/src/main/java/ru/surfstudio/android/core/ui/base/image/ImageLoader.java
package ru.surfstudio.android.core.ui.base.image;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.VectorDrawable;
import android.support.annotation.DrawableRes;
import android.support.annotation.WorkerThread;
import android.support.graphics.drawable.VectorDrawableCompat;
import android.support.v4.content.ContextCompat;
import android.view.View;
import android.widget.ImageView;

import com.bumptech.glide.BitmapRequestBuilder;
import com.bumptech.glide.Glide;
import com.bumptech.glide.load.Transformation;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.load.resource.bitmap.CenterCrop;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.animation.GlideAnimation;
import com.bumptech.glide.request.target.SimpleTarget;
import com.bumptech.glide.request.target.Target;
import com.bumptech.glide.request.target.ViewTarget;

import java.io.File;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutionException;

import ru.surfstudio.android.core.R;
import ru.surfstudio.android.core.ui.base.image.RoundedCornersTransformation.CornerType;
import ru.surfstudio.android.core.ui.base.image.blur.BlurTransformation;
import ru.surfstudio.android.logger.Logger;

/**
 * Класс-прокси, реализующий загрузку изображений с необходимостью абстрагироваться от реализации контракта библиотеки
 * Необходимо проксировать абсолютно все запросы на отображение изображения
 */
public final class ImageLoader {

    public static final int NO_SIZE = -1;

    private Context context;
    private View view;
    private String url;
    @DrawableRes
    private int drawableUri;
    @DrawableRes
    private int previewResId;
    @DrawableRes
    private int errorResId;

    private int maxWidth = NO_SIZE;
    private int maxHeight = NO_SIZE;
    private int cornerRadius;
    private int cornerMargin;
    private CornerType cornerType;
    private int blurRadius = 20;
    private int blurSampling = 1;

    private boolean centerCrop;

    private boolean circle;

    @DrawableRes
    private int overlayResId;

    @DrawableRes
    private int maskResId;

    private boolean skipCache;
    private boolean blurTransform;

    private boolean filterBitmapOnScale = false;
    private OnImageLoadedListener listener;
    private OnImageErrorLoadListener errorLoadLisner;
    private RequestListener<Serializable, Bitmap> glideListener = new RequestListener<Serializable, Bitmap>() {
        @Override
        public boolean onException(Exception e, Serializable model, Target<Bitmap> target, boolean isFirstResource) {
            if (view != null) {
                view.setTag(R.id.image_loader_tag, null);
            }

            if (errorLoadLisner != null) {
                errorLoadLisner.onLoadError(e);
            }
            return false;
        }

        @Override
        public boolean onResourceReady(Bitmap resource,
                                       Serializable model,
                                       Target<Bitmap> target,
                                       boolean isFromMemoryCache,
                                       boolean isFirstResource) {
            if (listener != null) {
                listener.onLoadFinish(resource);
            }
            return false;
        }
    };

    private ImageLoader(Context context) {
        this.context = context;
    }

    public static ImageLoader with(Context context) {
        return new ImageLoader(context);
    }

    private static Bitmap getBitmap(VectorDrawableCompat vectorDrawable) {
        Bitmap bitmap = Bitmap.createBitmap(vectorDrawable.getIntrinsicWidth(),
                vectorDrawable.getIntrinsicHeight(), Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        vectorDrawable.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
        vectorDrawable.draw(canvas);
        return bitmap;
    }

    private static Bitmap getBitmap(VectorDrawable vectorDrawable) {
        Bitmap bitmap = Bitmap.createBitmap(vectorDrawable.getIntrinsicWidth(),
                vectorDrawable.getIntrinsicHeight(), Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        vectorDrawable.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
        vectorDrawable.draw(canvas);
        return bitmap;
    }

    public ImageLoader url(String url) {
        this.url = url;
        return this;
    }

    public ImageLoader url(@DrawableRes int drawableUri) {
        this.drawableUri = drawableUri;
        return this;
    }

    public ImageLoader centerCrop(boolean centerCrop) {
        this.centerCrop = centerCrop;
        return this;
    }

    public ImageLoader circle(boolean circle) {
        this.circle = circle;
        return this;
    }

    /**
     * устанавливает максимальный размер изображения
     * отношение ширины и высоты останется неизменным
     */
    public ImageLoader maxWidth(int maxWidth) {
        this.maxWidth = maxWidth;
        return this;
    }

    /**
     * устанавливает максимальный размер изображения
     * отношение ширины и высоты останется неизменным
     */
    public ImageLoader maxHeight(int maxHeight) {
        this.maxHeight = maxHeight;
        return this;
    }

    /**
     * радиус для обрезания углов с отступом
     *
     * @param radius радиус угла в пикселях
     * @param margin отступ от края в пикселях
     */
    public ImageLoader roundedCorners(int radius, int margin, CornerType cornerType) {
        this.cornerRadius = radius;
        this.cornerMargin = margin;
        this.cornerType = cornerType;
        return this;
    }

    public ImageLoader roundedAllCorners(int radius, int margin) {
        this.cornerRadius = radius;
        this.cornerMargin = margin;
        this.cornerType = CornerType.ALL;
        return this;
    }

    public ImageLoader blur(boolean blurTransform) {
        this.blurTransform = blurTransform;
        return this;
    }

    public ImageLoader blur(int radius, int sampling) {
        blurTransform = true;
        blurRadius = radius;
        blurSampling = sampling;
        return this;
    }

    /**
     * Наложение drawable с {@code PorterDuff.Mode.DST_ATOP}
     */
    public ImageLoader overlay(@DrawableRes int resId) {
        overlayResId = resId;
        return this;
    }

    /**
     * Масковая трансформация, может работать с NinePatch
     */
    public ImageLoader mask(@DrawableRes int resId) {
        maskResId = resId;
        return this;
    }

    public ImageLoader preview(@DrawableRes int resId) {
        this.previewResId = resId;
        return this;
    }

    public ImageLoader error(@DrawableRes int resId) {
        this.errorResId = resId;
        return this;
    }

    public ImageLoader skipCache(boolean skipCache) {
        this.skipCache = skipCache;
        return this;
    }

    /**
     * Фильтровать ли изображения при масштабировании
     *
     * @see Bitmap#createScaledBitmap(Bitmap, int, int, boolean)
     */
    public ImageLoader filterBitMapOnScale(boolean filterBitmapOnScale) {
        this.filterBitmapOnScale = filterBitmapOnScale;
        return this;
    }

    @WorkerThread
    public Bitmap get() {
        Bitmap result;
        try {
            result = Glide.with(context)
                    .load(url)
                    .asBitmap()
                    .into(NO_SIZE, NO_SIZE)
                    .get();
        } catch (InterruptedException | ExecutionException e) {
            if (errorResId == 0) {
                throw new IllegalStateException("Error load bitmap", e);
            }
            if (errorLoadLisner != null) {
                errorLoadLisner.onLoadError(e);
            }
            result = getFromResource(errorResId);
        }
        return result;
    }

    public void downloadOnly() {
        Glide.with(context)
                .load(url)
                .downloadOnly(new SimpleTarget<File>() {
                    @Override
                    public void onResourceReady(File resource, GlideAnimation<? super File> glideAnimation) {
                        Logger.d("Image loaded to cache " + (resource != null ? resource.getAbsolutePath() : "null" ));
                    }

                    @Override
                    public void onLoadFailed(Exception e, Drawable errorDrawable) {
                        super.onLoadFailed(e, errorDrawable);
                        if (errorLoadLisner != null) {
                            errorLoadLisner.onLoadError(e);
                        }
                    }
                });
    }

    public void in(ImageView imageView) {
        view = imageView;

        if (imageView.getTag(R.id.image_loader_tag) != null &&
                imageView.getTag(R.id.image_loader_tag).equals(url)) {
            return;
        }

        imageView.setTag(R.id.image_loader_tag, url);
        bitmapRequestBuilder().into(imageView);
    }

    public void in(View view) {
        this.view = view;

        if (view.getTag(R.id.image_loader_tag) != null &&
                view.getTag(R.id.image_loader_tag).equals(url)) {
            return;
        }

        view.setTag(R.id.image_loader_tag, url);
        bitmapRequestBuilder().into(new ViewTarget<View, Bitmap>(view) {
            @Override
            public void onLoadFailed(Exception e, Drawable errorDrawable) {
                super.onLoadFailed(e, errorDrawable);
                view.setBackground(getCompatDrawable(errorResId));
            }

            @Override
            public void onResourceReady(Bitmap resource, GlideAnimation glideAnimation) {
                view.setBackground(new BitmapDrawable(context.getResources(), resource));
            }
        });
    }

    public void in(SimpleTarget simpleTarget) {
        bitmapRequestBuilder().into(new SimpleTarget<Bitmap>() {
            @Override
            public void onLoadFailed(Exception e, Drawable errorDrawable) {
                super.onLoadFailed(e, errorDrawable);
                simpleTarget.onResourceReady(getFromResource(errorResId), null);

            }

            @Override
            public void onResourceReady(Bitmap resource, GlideAnimation<? super Bitmap> glideAnimation) {
                simpleTarget.onResourceReady(resource, glideAnimation);
            }
        });
    }

    public ImageLoader clear(ImageView imageView) {
        Glide.clear(imageView);
        return this;
    }

    private Drawable getCompatDrawable(int previewResId) {
        return previewResId != 0
                ? ContextCompat.getDrawable(context, previewResId)
                : null;
    }

    public ImageLoader listener(OnImageLoadedListener listener) {
        this.listener = listener;
        return this;
    }

    public ImageLoader errorlistener(OnImageErrorLoadListener errorLoadListener) {
        this.errorLoadLisner = errorLoadListener;
        return this;
    }

    @WorkerThread
    private Bitmap getFromResource(int drawableId) {
        Drawable drawable = ContextCompat.getDrawable(context, drawableId);
        if (drawable instanceof BitmapDrawable) {
            return ((BitmapDrawable) drawable).getBitmap();
        } else if (drawable instanceof VectorDrawableCompat) {
            return getBitmap((VectorDrawableCompat) drawable);
        } else if (drawable instanceof VectorDrawable) {
            //noinspection NewApi
            return getBitmap((VectorDrawable) drawable);
        } else {
            throw new IllegalArgumentException("unsupported drawable type " + drawable.getClass().getSimpleName());
        }
    }

    private BitmapRequestBuilder bitmapRequestBuilder() {
        List<Transformation> transforms = new ArrayList<>();
        transforms.add(new SizeTransformation(context,
                maxWidth, maxHeight, filterBitmapOnScale));

        if (centerCrop) {
            transforms.add(new CenterCrop(context));
        }

        if (circle) {
            transforms.add(new CircleCropTransformation(context));
        }

        if (cornerRadius > 0 || cornerMargin > 0) {
            transforms.add(new RoundedCornersTransformation(context,
                    cornerRadius, cornerMargin, cornerType));
        }

        if (blurTransform) {
            transforms.add(new BlurTransformation(context, blurRadius, blurSampling));
        }

        if (overlayResId > 0) {
            transforms.add(new OverlayTransformation(context, overlayResId));
        }

        // масковая трансформация должна накладываться последней
        // для обрезания конечного результата
        if (maskResId > 0) {
            transforms.add(new MaskTransformation(context, maskResId));
        }

        return Glide.with(context)
                .load(drawableUri != 0 ? drawableUri : url)
                .asBitmap()
                .skipMemoryCache(skipCache)
                .diskCacheStrategy(skipCache ? DiskCacheStrategy.NONE : DiskCacheStrategy.ALL)
                .placeholder(getCompatDrawable(previewResId))
                .error(getCompatDrawable(errorResId))
                .animate(R.anim.fade_in) // workaround https://github.com/bumptech/glide/issues/363
                .listener(glideListener)
                .transform(transforms.toArray(new Transformation[transforms.size()]));
    }

    public interface OnImageLoadedListener {
        void onLoadFinish(Bitmap bitmap);
    }

    public interface OnImageErrorLoadListener {
        void onLoadError(Throwable throwable);
    }
}

File: imageloader/src/main/java/ru/surfstudio/android/imageloader/ImageLoader.kt
/*
  Copyright (c) 2018-present, SurfStudio LLC.

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
package ru.surfstudio.android.imageloader

import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.support.annotation.DrawableRes
import android.support.annotation.WorkerThread
import android.view.View
import android.widget.ImageView
import com.bumptech.glide.Glide
import com.bumptech.glide.RequestBuilder
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.load.resource.bitmap.DownsampleStrategy
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.SimpleTarget
import com.bumptech.glide.request.target.Target
import com.bumptech.glide.request.target.ViewTarget
import com.bumptech.glide.request.transition.Transition
import ru.surfstudio.android.imageloader.data.*
import ru.surfstudio.android.imageloader.transformations.BlurTransformation.BlurBundle
import ru.surfstudio.android.imageloader.transformations.MaskTransformation.OverlayBundle
import ru.surfstudio.android.imageloader.transformations.RoundedCornersTransformation.CornerType
import ru.surfstudio.android.imageloader.transformations.RoundedCornersTransformation.RoundedCornersBundle
import ru.surfstudio.android.logger.Logger
import ru.surfstudio.android.utilktx.util.DrawableUtil
import java.util.concurrent.ExecutionException

/**
 * Загрузчик изображений.
 *
 * Является надстройкой над библиотекой Glide.
 */
class ImageLoader(private val context: Context) : ImageLoaderInterface {

    private var imageCacheManager: ImageCacheManager = ImageCacheManager()
    private var imageTransformationsManager: ImageTransformationsManager =
            ImageTransformationsManager(context)
    private var imageResourceManager: ImageResourceManager =
            ImageResourceManager(context, imageTransformationsManager)
    private var imageTargetManager: ImageTargetManager =
            ImageTargetManager(context, imageResourceManager)
    private var imageTagManager: ImageTagManager =
            ImageTagManager(imageTargetManager, imageResourceManager)

    private var onImageLoadedLambda: ((drawable: Drawable) -> (Unit))? = null
    private var onImageLoadErrorLambda: ((throwable: Throwable) -> (Unit))? = null

    private val glideDownloadListener = object : RequestListener<Drawable> {
        override fun onLoadFailed(e: GlideException?,
                                  model: Any?,
                                  target: Target<Drawable>?,
                                  isFirstResource: Boolean) = false.apply {
            imageTagManager.setTag(null)
            e?.let { onImageLoadErrorLambda?.invoke(it) }
        }

        override fun onResourceReady(resource: Drawable,
                                     model: Any?,
                                     target: Target<Drawable>?,
                                     dataSource: DataSource?,
                                     isFirstResource: Boolean): Boolean =
                false.apply { onImageLoadedLambda?.invoke(resource) }
    }

    companion object {
        fun with(context: Context) = ImageLoader(context)
    }

    /**
     * Загрузка изображения из сети
     *
     * @param url сетевая ссылка на изображение
     */
    @Throws(IllegalArgumentException::class)
    override fun url(url: String) =
            apply {
                this.imageResourceManager.url = url
            }

    /**
     * Загрузка изображения из ресурсов
     *
     * @param drawableResId ссылка на ресурс из папки res/drawable
     */
    override fun url(@DrawableRes drawableResId: Int) =
            apply { this.imageResourceManager.drawableResId = drawableResId }

    /**
     * Указание графического ресурса, отображаемого в качестве плейсхолдера
     *
     * @param drawableResId ссылка на ресурс из папки res/drawable
     */
    override fun preview(@DrawableRes drawableResId: Int) =
            apply { this.imageResourceManager.previewResId = drawableResId }

    /**
     * Указание графического ресурса, отображаемого в случае ошибки загрузки
     *
     * @param drawableResId ссылка на ресурс из папки res/drawable
     */
    override fun error(@DrawableRes drawableResId: Int) =
            apply { this.imageResourceManager.errorResId = drawableResId }

    /**
     * Установка лямбды для отслеживания загрузки изображения
     *
     * @param lambda лямбда, возвращающая загруженный [Drawable]
     */
    override fun listener(lambda: ((drawable: Drawable) -> (Unit))) =
            apply { this.onImageLoadedLambda = lambda }

    /**
     * Установка лямбды для отслеживания ошибки при загрузке изображения
     *
     * @param lambda лямбда, возвращающая ошибку [Throwable]
     */
    override fun errorListener(lambda: ((throwable: Throwable) -> (Unit))) =
            apply { this.onImageLoadErrorLambda = lambda }

    /**
     * Указание политики кэширования.
     * Метод предоставляет возможность отключить кэширование загруженных изображений в памяти и на диске.
     *
     * В данной реализации кэшируется только финальное изображение (прошедшее необходимую постобработку и трансформации).
     * @param skipCache true - игнорировать кэш в памяти и на диске, false - использовать кэш в памяти и на диске
     */
    override fun skipCache(skipCache: Boolean) =
            apply { this.imageCacheManager.skipCache = skipCache }

    /**
     * Установка максимальной ширины изображения в px.
     *
     * Необходима для пережатия изображения без искажения пропорций.
     */
    override fun maxWidth(maxWidth: Int) =
            apply { this.imageTransformationsManager.imageSizeManager.maxWidth = maxWidth }

    /**
     * Установка максимальной высоты изображения в px.
     *
     * Необходима для пережатия изображения без искажения пропорций.
     */
    override fun maxHeight(maxHeight: Int) =
            apply { this.imageTransformationsManager.imageSizeManager.maxHeight = maxHeight }

    /**
     * Масштабирование изображения по размеру виджета с обрезкой излишков.
     *
     * @param isCrop флаг активации трансформации
     */
    override fun centerCrop(isCrop: Boolean) =
            apply { this.imageTransformationsManager.isCenterCrop = isCrop }

    /**
     * Преобразование прямоугольного изображения в круглое.
     *
     * @param isCircle флаг активации трансформации
     */
    override fun circle(isCircle: Boolean) =
            apply { imageTransformationsManager.isCircle = isCircle }

    /**
     * Скругление углов у прямоугольного изображения.
     *
     * @param isRoundedCorners флаг активации трансформации
     * @param radiusPx радиус скругления в px
     * @param marginPx величина отступа в px
     * @param cornerType конфигурация скругляемых углов
     */
    override fun roundedCorners(isRoundedCorners: Boolean, radiusPx: Int, marginPx: Int, cornerType: CornerType) =
            also {
                imageTransformationsManager.roundedCornersBundle =
                        RoundedCornersBundle(isRoundedCorners, cornerType, radiusPx, marginPx)
            }

    /**
     * Эффект размытия изображения "Blur".
     *
     * @param isBlur флаг активации трансформации
     * @param blurRadiusPx радиус размытия
     * @param blurDownSampling уровень принудительного понижения качества разрешения изображения
     */
    override fun blur(isBlur: Boolean, blurRadiusPx: Int, blurDownSampling: Int) =
            also {
                imageTransformationsManager.blurBundle =
                        BlurBundle(isBlur, blurRadiusPx, blurDownSampling)
            }

    /**
     * Наложение маски на изображение с поддержкой 9-patch маски.
     *
     * @param isOverlay флаг активации трансформации
     * @param maskResId ссылка на ресурс изображения маски из папки res/drawable
     */
    override fun mask(isOverlay: Boolean, @DrawableRes maskResId: Int) =
            also {
                imageTransformationsManager.overlayBundle = OverlayBundle(isOverlay, maskResId)
            }

    override fun sizeMultiplier(sizeMultiplier: Float): ImageLoaderInterface =
            also {
                imageTransformationsManager.sizeMultiplier = sizeMultiplier
            }

    /**
     * Указание целевой [View]
     *
     * @param view экземпляр [View] для загрузки изображения
     */
    override fun into(view: View) {
        this.imageTargetManager.targetView = view

        if (imageTagManager.isTagUsed()) return

        imageTagManager.setTag(imageResourceManager.url)

        performLoad(view)
    }

    /**
     * Загрузка изображения в [SimpleTarget].
     *
     * @param simpleTarget обработчик загрузки изображения.
     */
    fun into(simpleTarget: SimpleTarget<Drawable>) {
        into(
                { resource, transition -> simpleTarget.onResourceReady(resource, transition) },
                {
                    it?.let {
                        simpleTarget.onResourceReady(it, null)
                    }
                }
        )
    }

    /**
     * Загрузка изображения в [SimpleTarget]
     *
     * @param resourceReadyLambda колбек в случае успеха
     * @param loadFailedLambda колбек при ошибке
     */
    fun into(
            resourceReadyLambda: (resource: Drawable, transition: Transition<in Drawable>?) -> Unit,
            loadFailedLambda: (errorDrawable: Drawable?) -> Unit
    ) {
        buildRequest().into(object : SimpleTarget<Drawable>() {
            override fun onResourceReady(resource: Drawable, transition: Transition<in Drawable>?) {
                resourceReadyLambda(resource, transition)
            }

            override fun onLoadFailed(errorDrawable: Drawable?) {
                super.onLoadFailed(errorDrawable)
                loadFailedLambda(errorDrawable)
            }
        })
    }

    /**
     * Получение исходника изображения в формате [Bitmap].
     * Кейс использования - загрузка изображения на уровне интерактора для отправки на сервер.
     * Без отображения на UI.
     * Для отображения на UI использовать [into]
     *
     * Запрос происходит в UI-потоке.
     */
    @WorkerThread
    override fun get(): Bitmap? {
        var result: Bitmap?
        try {
            result = Glide.with(context)
                    .asBitmap()
                    .load(imageResourceManager.toLoad())
                    .apply(
                            RequestOptions()
                                    .diskCacheStrategy(if (imageCacheManager.skipCache) DiskCacheStrategy.NONE else DiskCacheStrategy.ALL)
                                    .skipMemoryCache(imageCacheManager.skipCache)
                                    .transforms(*imageTransformationsManager.prepareTransformations())
                    )
                    .submit()
                    .get()
        } catch (e: Exception) {
            when (e) {
                is InterruptedException, is ExecutionException -> {
                    if (!imageResourceManager.isErrorPresented()) {
                        Logger.e("ImageLoader.get() / Ошибка загрузки изображения")
                        throw IllegalStateException("Ошибка загрузки изображения")
                    }
                    onImageLoadErrorLambda?.invoke(e)
                    result = DrawableUtil.getBitmapFromRes(context, imageResourceManager.errorResId)
                }
                else -> throw e
            }
        }
        return result
    }


    /**
     * Формирование запроса на загрузку изображения.
     */
    private fun buildRequest(): RequestBuilder<Drawable> = Glide.with(context)
            .load(imageResourceManager.toLoad())
            .error(imageResourceManager.prepareErrorDrawable())
            .thumbnail(imageResourceManager.preparePreviewDrawable())
            .apply(
                    RequestOptions()
                            /*.diskCacheStrategy(if (imageCacheManager.skipCache) DiskCacheStrategy.NONE else DiskCacheStrategy.ALL)
                            .skipMemoryCache(imageCacheManager.skipCache)
                            .downsample(DownsampleStrategy.AT_MOST)
                            .sizeMultiplier(imageTransformationsManager.sizeMultiplier)*/
                            .transforms(*imageTransformationsManager.prepareTransformations())
            )
            .listener(glideDownloadListener)

    /**
     * Загрузка изображения в целевую [View].
     *
     * @param view целевая [View]
     */
    private fun performLoad(view: View) {
        if (view is ImageView) {
            buildRequest().into(view)
        } else {
            buildRequest().into(object : ViewTarget<View, Drawable>(view) {

                override fun onLoadStarted(placeholder: Drawable?) {
                    super.onLoadStarted(placeholder)
                    view.background = placeholder
                }

                override fun onResourceReady(resource: Drawable, transition: Transition<in Drawable>?) {
                    view.background = resource
                }

                override fun onLoadFailed(errorDrawable: Drawable?) {
                    super.onLoadFailed(errorDrawable)
                    view.background = errorDrawable
                }
            })
        }
    }

}